src/types/dict.rs

```diff
@@ -372,85 +396,333 @@ fn dict_len(dict: &Bound<'_, PyDict>) -> Py_ssize_t {
+            #[inline]
+            fn check<T>(
+                mut predicate: impl FnMut(&T) -> bool,
+            ) -> impl FnMut((), T) -> ControlFlow<T> {
+                move |(), x| {
+                    if predicate(&x) {
+                        ControlFlow::Break(x)
+                    } else {
+                        ControlFlow::Continue(())
+                    }
+                }
+            }
+
+            match iter.try_fold((), check(predicate)) {
+                ControlFlow::Continue(_) => None,
+                ControlFlow::Break(x) => Some(x),
+            }
+        })
+    }
+
+    #[inline]
+    #[cfg(all(Py_GIL_DISABLED, not(feature = "nightly")))]
+    fn find_map<B, F>(&mut self, f: F) -> Option<B>
+    where
+        Self: Sized,
+        F: FnMut(Self::Item) -> Option<B>,
+    {
+        self.with_critical_section(|iter| {
+            #[inline]
+            fn check<T, B>(
+                mut f: impl FnMut(T) -> Option<B>,
+            ) -> impl FnMut((), T) -> ControlFlow<B> {
+                move |(), x| match f(x) {
+                    Some(x) => ControlFlow::Break(x),
+                    None => ControlFlow::Continue(()),
+                }
+            }
+
+            match iter.try_fold((), check(f)) {
+                ControlFlow::Continue(_) => None,
+                ControlFlow::Break(x) => Some(x),
+            }
+        })
+    }
+
+    #[inline]
+    #[cfg(all(Py_GIL_DISABLED, not(feature = "nightly")))]
+    fn position<P>(&mut self, predicate: P) -> Option<usize>
+    where
+        Self: Sized,
+        P: FnMut(Self::Item) -> bool,
+    {
+        self.with_critical_section(|iter| {
+            #[inline]
+            fn check<'a, T>(
+                mut predicate: impl FnMut(T) -> bool + 'a,
+                acc: &'a mut usize,
+            ) -> impl FnMut((), T) -> ControlFlow<usize, ()> + 'a {
+                move |_, x| {
+                    if predicate(x) {
+                        ControlFlow::Break(*acc)
+                    } else {
+                        *acc += 1;
+                        ControlFlow::Continue(())
+                    }
+                }
+            }
+
+            let mut acc = 0;
+            match iter.try_fold((), check(predicate, &mut acc)) {
+                ControlFlow::Continue(_) => None,
+                ControlFlow::Break(x) => Some(x),
+            }
+        })
+    }
+}
+
+#[inline]
+#[cfg(Py_GIL_DISABLED)]
+pub fn with_critical_section<F, R>(object: &Bound<'_, PyAny>, f: F) -> R
+where
+    F: FnOnce() -> R,
+{
+    struct Guard(ffi::PyCriticalSection);
+
+    impl Drop for Guard {
+        fn drop(&mut self) {
+            unsafe {
+                ffi::PyCriticalSection_End(&mut self.0);
+            }
+        }
+    }
+
+    let _guard = unsafe {
+        let mut section = std::mem::zeroed();
+        ffi::PyCriticalSection_Begin(&mut section, object.as_ptr());
+        Guard(section)
+    };
+    f()
 }
```

Comment:
I think @ngoldbaum and I came up with pretty much the same API, I think probably this belongs in `src/sync.rs`?